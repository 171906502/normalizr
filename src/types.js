// @flow
import ArraySchema from './schemas/Array';
import EntitySchema from './schemas/Entity';
import ObjectSchema from './schemas/Object';
import UnionSchema from './schemas/Union';
import ValuesSchema from './schemas/Values';

export type Schema = Array<Schema> | {} | ArraySchema | EntitySchema | ObjectSchema | UnionSchema | ValuesSchema;

export type Visitor = (value: *, parent: ?{}, key: ?string, schema: Schema, addEntity: AddEntity) => *;

export type AddEntity = (schema: EntitySchema, processedEntity: {}, value: {}, parent: *, key: ?string) => void;

export type Unvisitor = (input: *, schema: Schema) => mixed;
