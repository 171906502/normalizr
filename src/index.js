import EntitySchema from './schemas/Entity';
import UnionSchema from './schemas/Union';
import ValuesSchema from './schemas/Values';
import ArraySchema, { normalize as normalizeArray } from './schemas/Array';
import ObjectSchema, { normalize as normalizeObject } from './schemas/Object';

const visit = (value, parent, key, schema, addEntity) => {
  if (typeof value !== 'object' || !value) {
    return value;
  }

  if (typeof schema === 'object' && (!schema.normalize || typeof schema.normalize !== 'function')) {
    let method = normalizeObject;
    if (Array.isArray(schema)) {
      method = normalizeArray;
    }
    return method(schema, value, parent, key, visit, addEntity);
  }

  return schema.normalize(value, parent, key, visit, addEntity);
};

const addEntities = (entities) => (schema, processedEntity, value, parent, key) => {
  const schemaKey = schema.getKey(value, parent, key);
  const id = schema.getId(value, parent, key);
  if (!(schemaKey in entities)) {
    entities[schemaKey] = {};
  }

  const existingEntity = entities[schemaKey][id];
  if (existingEntity) {
    entities[schemaKey][id] = schema.merge(existingEntity, processedEntity);
  } else {
    entities[schemaKey][id] = processedEntity;
  }
};

export const schema = {
  Array: ArraySchema,
  Entity: EntitySchema,
  Object: ObjectSchema,
  Union: UnionSchema,
  Values: ValuesSchema
};

export const normalize = (input, schema) => {
  if (!input || typeof input !== 'object') {
    throw new Error(`Unexpected input given to normalize. Expected type to be "object", found "${typeof input}".`);
  }

  const entities = {};
  const addEntity = addEntities(entities);

  const result = visit(input, input, null, schema, addEntity);
  return { entities, result };
};
