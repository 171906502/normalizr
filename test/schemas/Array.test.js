/* eslint-env jest */
import { normalize, schema } from '../../src';

describe(schema.Array.name, () => {
  it('normalizes an array of objects', () => {
    const entity = new schema.Entity('item');
    const arraySchema = new schema.Array(entity);
    expect(normalize([ { id: 1 }, { id: 2 } ], arraySchema)).toMatchSnapshot();
  });

  it(`auto-converts plain arrays to ${schema.Array.name}`, () => {
    const userSchema = new schema.Entity('user');
    expect(normalize([ { id: 1 }, { id: 2 } ], [ userSchema ])).toMatchSnapshot();
  });

  it('throws an error if created with more than one schema', () => {
    const userSchema = new schema.Entity('users');
    const catSchema = new schema.Entity('cats');
    expect(() => normalize([ { id: 1 } ], [ catSchema, userSchema ])).toThrow();
    expect(() => normalize([ { id: 1 } ], new schema.Array([ catSchema, userSchema ]))).toThrow();
  });
});
