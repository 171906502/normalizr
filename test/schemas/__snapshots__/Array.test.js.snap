exports[`ArraySchema auto-converts plain arrays to ArraySchema 1`] = `
Object {
  "entities": Object {
    "user": Object {
      "1": Object {
        "id": 1,
      },
      "2": Object {
        "id": 2,
      },
    },
  },
  "result": Array [
    1,
    2,
  ],
}
`;

exports[`ArraySchema normalizes an array of objects 1`] = `
Object {
  "entities": Object {
    "item": Object {
      "1": Object {
        "id": 1,
      },
      "2": Object {
        "id": 2,
      },
    },
  },
  "result": Array [
    1,
    2,
  ],
}
`;
